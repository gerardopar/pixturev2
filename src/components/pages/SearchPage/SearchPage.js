// importing modules
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import queryString from 'query-string';
// importing redux actions
import {
 setImagesAsync, setOpenModal, 
        setCloseModal, setImagesSearchedAsync, setPaginationAsync 
} from '../../../actions/searchPageActions';

// importing components
import SearchHeader from './SearchHeader/SearchHeader';
import ImageList from '../../UI/ImageList/ImageList';
import ImageModal from '../../UI/ImageModal/ImageModal';
import Pagination from '../../UI/pagination/Pagination';
import MobileSearchBar from '../../UI/mobile/MobileSearchBar';
import Spinner from '../../UI/spinner/Spinner';

class SearchPage extends Component {
    constructor(props) {
        super(props);
        this.state = {
            page: 1,
            tag: '' 
        };
    }

    componentDidMount() {
        const imageSearched = this.props.match.params.image;
        const parsedPageQuery = queryString.parse(this.props.location.search);

        this.setState({ page: Number(parsedPageQuery.page), tag: imageSearched });
        this.handleImageFetch(imageSearched);
    }

    handleImageFetch = (query) => {
        const parsedPageQuery = queryString.parse(this.props.location.search);
        this.props.setImagesAsync(query, parsedPageQuery);
    }

    handleImageSearch = (e) => {
        e.preventDefault();
        const imageSearch = e.target.elements.imageSearched.value; // user input
        const parsedPageQuery = queryString.parse(this.props.location.search); // query params

        this.props.setImagesSearchedAsync(imageSearch, parsedPageQuery);
        this.setState({ page: 1, tag: imageSearch });
        this.props.history.push(`/search/${imageSearch}?page=1`);

        document.getElementById('searchPage__form').reset();
        document.getElementById('mobileSearchBar__form').reset();
    }

    handleOpenModal = (imgUrl, height, width, user, views, likes) => {
        this.props.setOpenModal(imgUrl, height, width, user, views, likes);
    }

    handleCloseModal = () => {
        this.props.setCloseModal();
    }

    handlePagination = (direction) => {  
        const imageSearched = this.props.match.params.image;
        const parsedPageQuery = queryString.parse(this.props.location.search);

        if (direction === 'next') {
            this.setState(prevState => ({
                    page: Number(prevState.page) + 1
                }));
        } else if (direction === 'previous') {
            this.setState(prevState => ({
                    page: Number(prevState.page) - 1
                }));
        }

        parsedPageQuery.page = this.state.page;
        this.props.history.push(`/search/${imageSearched}?page=${this.state.page}`);
        this.props.setPaginationAsync(imageSearched, parsedPageQuery);
        window.scrollTo(0, 0);
    }

    render() {
        return (
            <div className="searchPage">
                <SearchHeader handleImageSearch={this.handleImageSearch} />
                <MobileSearchBar handleImageSearch={this.handleImageSearch} />
                <div className="tag">
                    <div className="tag__wrap">
                        <p className="tag__title">
                            Free images of
                            {' '}
                            { this.state.tag }
                        </p>
                    </div>
                </div>
                {
                    this.props.images.length > 0 
                    ? (
                <ImageList 
                  handleOpenModal={this.handleOpenModal}
                  images={this.props.images}
                />
                ) : <Spinner />
                }
                {
                    this.props.modal.hidden 
                    ? null : (
                    <ImageModal 
                      {...this.props.modal}
                      handleCloseModal={this.handleCloseModal} 
                    />
                    )
                }
                <Pagination 
                  handlePagination={this.handlePagination}
                  page={this.state.page}
                />
            </div>
        );
    }
}

// # redux state
const mapStateToProps = state => ({
    images: state.home.images,
    modal: state.home.modal
});

// # redux actions
const mapDispatchToProps = dispatch => ({
    setImagesAsync: (query, parsedPageQuery) => dispatch(setImagesAsync(query, parsedPageQuery)),
    setImagesSearchedAsync: (query, parsedPageQuery) => dispatch(setImagesSearchedAsync(query, parsedPageQuery)),
    setOpenModal: (imgUrl, height, width, user, views, likes) => dispatch(setOpenModal(imgUrl, height, width, user, views, likes)),
    setCloseModal: () => dispatch(setCloseModal()),
    setPaginationAsync: (query, parsedPageQuery) => dispatch(setPaginationAsync(query, parsedPageQuery))
});

SearchPage.propTypes = {
    history: PropTypes.objectOf(PropTypes.any),
    images: PropTypes.arrayOf(PropTypes.object),
    location: PropTypes.objectOf(PropTypes.any),
    match: PropTypes.objectOf(PropTypes.any),
    modal: PropTypes.shape({
        img: PropTypes.string,
        height: PropTypes.number,
        hidden: PropTypes.bool,
        likes: PropTypes.number,
        tags: PropTypes.string,
        user: PropTypes.string,
        views: PropTypes.number,
        width: PropTypes.number
    }),
    setImagesAsync: PropTypes.func,
    setImagesSearchedAsync: PropTypes.func,
    setOpenModal: PropTypes.func,
    setPaginationAsync: PropTypes.func,
    setCloseModal: PropTypes.func
};

SearchPage.defaultProps = {
    history: {},
    images: [],
    location: {},
    match: {},
    modal: {},
    setImagesAsync: () => {},
    setImagesSearchedAsync: () => {},
    setOpenModal: () => {},
    setPaginationAsync: () => {},
    setCloseModal: () => {}
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchPage);
