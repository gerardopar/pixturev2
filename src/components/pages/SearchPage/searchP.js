// importing modules
import React, { Component } from 'react';
import { connect } from 'react-redux';
import queryString from 'query-string';

// importing components
import SearchHeader from '../SearchPage/SearchHeader/SearchHeader';
import ImageList from '../../UI/ImageList/ImageList';
import ImageModal from '../../UI/ImageModal/ImageModal';
import Pagination from '../../UI/pagination/Pagination';
import MobileSearchBar from '../../UI/mobile/MobileSearchBar';
import Spinner from '../../UI/spinner/Spinner';

class SearchPage extends Component {
    constructor(props){
        super(props);
        this.state = {
            images: [],
            modal: {
                img: '',
                height: 0,
                hidden: true,
                likes: 0,
                tags: '',
                user: '',
                views: 0,
                width: 0
            },
            page: 1,
            tag: ''
        }
    }

    componentDidMount(){
        const imageSearched = this.props.match.params.image;
        const parsedPageQuery = queryString.parse(this.props.location.search);
        this.setState({ page: Number(parsedPageQuery.page), tag: imageSearched });
        this.handleImageFetch(imageSearched);
    }

    handleImageFetch = (query) => {
        const parsedPageQuery = queryString.parse(this.props.location.search);

        fetch(`https://pixture.herokuapp.com/postImageQuery`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                imageSearched: query,
                page: parsedPageQuery.page
            })
        })
        .then((data) => {
            return data.json();
        })
        .then((data) => {
            this.setState({ images: data.images });
        })
        .catch((err) => console.log(err));
    }

    handleImageSearch = (e) => {
        e.preventDefault();
        const imageSearch = e.target.elements.imageSearched.value; // user input
        const parsedPageQuery = queryString.parse(this.props.location.search); // query params

        fetch(`https://pixture.herokuapp.com/postImageQuery`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                imageSearched: imageSearch,
                page: parsedPageQuery.page
            })
        })
        .then((data) => {
            return data.json();
        })
        .then((data) => {
            this.setState({ images: data.images, page: 1, tag: imageSearch });
            this.props.history.push(`/search/${imageSearch}?page=1`);
            document.getElementById("searchPage__form").reset();
            document.getElementById("mobileSearchBar__form").reset();
        })
        .catch((err) => console.log(err));
    }

    handleOpenModal = (imgUrl, height, width, user, views, likes) => {
        this.setState({ modal : {
            hidden: false,
            img: imgUrl,
            height: height,
            likes: likes,
            user: user,
            views: views,
            width: width
        }});
    }

    handleCloseModal = () => {
        this.setState({
            modal: {
                hidden: true
            }
        })
    }

    handlePaginationNext = () => {  
        const imageSearched = this.props.match.params.image;
        const parsedPageQuery = queryString.parse(this.props.location.search);
        this.setState((prevState) => {
            return {
                page: Number(prevState.page) + 1
            }
        });
        parsedPageQuery.page = this.state.page;
        
        fetch(`https://pixture.herokuapp.com/postImageQuery`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                imageSearched: imageSearched,
                page: parsedPageQuery.page
            })
        })
        .then((data) => {
            return data.json();
        })
        .then((data) => {
            this.props.history.push(`/search/${imageSearched}?page=${this.state.page}`);
            return data.images;
        })
        .then((images) => {
            window.scrollTo(0, 0);
            this.setState({ images: images });
        })
        .catch((err) => console.log(err));
    }

    handlePaginationPrev = () => {  
        const imageSearched = this.props.match.params.image;
        const parsedPageQuery = queryString.parse(this.props.location.search);
        this.setState((prevState) => {
            return {
                page: Number(prevState.page) - 1
            }
        });
        parsedPageQuery.page = this.state.page;
        
        fetch(`https://pixture.herokuapp.com/postImageQuery`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                imageSearched: imageSearched,
                page: parsedPageQuery.page
            })
        })
        .then((data) => {
            return data.json();
        })
        .then((data) => {
            this.props.history.push(`/search/${imageSearched}?page=${this.state.page}`);
            return data.images;
        })
        .then((images) => {
            window.scrollTo(0, 0);
            this.setState({ images: images });
        })
        .catch((err) => console.log(err));
    }

    render(){
        return (
            <div className="searchPage">
                {/* searchpage header */}
                <SearchHeader handleImageSearch={this.handleImageSearch}/>
                <MobileSearchBar handleImageSearch={this.handleImageSearch}/>
                {/* searchpage content */}
                <div className="tag">
                    <div className="tag__wrap">
                        <p className="tag__title">Free images of { this.state.tag }</p>
                    </div>
                </div>
                {
                    this.state.images.length > 0 ? 
                    <ImageList 
                    handleOpenModal={this.handleOpenModal}
                    images={this.state.images}/> : <Spinner />
                }
                
                {
                    this.state.modal.hidden 
                    ? null : 
                    // image modal
                    <ImageModal 
                        {...this.state.modal}
                        handleCloseModal={this.handleCloseModal}/>
                }
                <Pagination 
                    handlePaginationNext={this.handlePaginationNext}
                    handlePaginationPrev={this.handlePaginationPrev}
                    page={this.state.page}
                />
            </div>
        );
    };
};

    // # redux state
    const mapStateToProps = (state) => {
        return {
            images: state.home.images,
            modal: state.home.modal
        }
    };

    // # redux actions
    const mapDispatchToProps = (dispatch) => {
        return {
            setImagesAsync: () => dispatch(setImagesAsync()),
            setOpenModal: (imgUrl, height, width, user, views, likes) => dispatch(setOpenModal(imgUrl, height, width, user, views, likes)),
            setCloseModal: () => dispatch(setCloseModal())
        }
    };

export default connect(mapStateToProps, mapDispatchToProps)(SearchPage);