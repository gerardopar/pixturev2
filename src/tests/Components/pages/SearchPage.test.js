
import React from 'react';
import { shallow } from 'enzyme';
import { SearchPage } from '../../../components/pages/SearchPage/SearchPage';

const searchPageProps = {
    modal: {
        img: 'https://imageTesturl.com',
        height: 100,
        hidden: false,
        likes: 1000,
        user: 'http://userTesturl.coom',
        views: 500,
        width: 200
    },
    images: [{ url: 'https://imageTesturl.com' }, { url: 'https://imageTesturl.com' }]
};

describe('renders SearchPage with expected props', () => {
    test('renders SearchPage with no props', () => {
        const wrapper = shallow(<SearchPage />);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    test('renders image modal with expected props', () => {
        const wrapper = shallow(<SearchPage modal={searchPageProps.modal} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    test('renders imageList with expected props ', () => {
        const wrapper = shallow(<SearchPage images={searchPageProps.images} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });
});
