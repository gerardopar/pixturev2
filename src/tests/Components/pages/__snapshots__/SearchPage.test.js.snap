// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SearchPage with expected props renders SearchPage with no props 1`] = `
"<div className=\\"searchPage\\">
  <searchHeader handleImageSearch={[Function]} />
  <mobileSearchBar handleImageSearch={[Function]} />
  <categoryTag tag=\\"\\" />
  <spinner />
  <imageModal handleCloseModal={[Function]} height={0} img=\\"\\" likes={0} user=\\"\\" views={0} width={0} />
  <pagination handlePagination={[Function]} page={1} />
</div>"
`;

exports[`renders SearchPage with expected props renders image modal with expected props 1`] = `
"<div className=\\"searchPage\\">
  <searchHeader handleImageSearch={[Function]} />
  <mobileSearchBar handleImageSearch={[Function]} />
  <categoryTag tag=\\"\\" />
  <spinner />
  <imageModal img=\\"https://imageTesturl.com\\" height={100} hidden={false} likes={1000} user=\\"http://userTesturl.coom\\" views={500} width={200} handleCloseModal={[Function]} />
  <pagination handlePagination={[Function]} page={1} />
</div>"
`;

exports[`renders SearchPage with expected props renders imageList with expected props  1`] = `
"<div className=\\"searchPage\\">
  <searchHeader handleImageSearch={[Function]} />
  <mobileSearchBar handleImageSearch={[Function]} />
  <categoryTag tag=\\"\\" />
  <imageList handleOpenModal={[Function]} images={{...}} />
  <imageModal handleCloseModal={[Function]} height={0} img=\\"\\" likes={0} user=\\"\\" views={0} width={0} />
  <pagination handlePagination={[Function]} page={1} />
</div>"
`;
